### PR TITLE
br: retry to scatter the regions if status is timeout or cancel

### LINKED ISSUE #47236: br need to retry scatter if reply status is not SUCCESS
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
br sends scatter request, and pd returns CANCEL response.

### 2. What did you expect to see? (Required)
br retries to send the request
### 3. What did you see instead (Required)
br ignores it.
### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by Leavrth**
### What problem does this PR solve?

Issue Number: close #47236

`ref #46001`

Problem Summary:
Due to the store limit, the regions with `CANCEL` or `TIMEOUT` status needs to be scatterred again.
### What is changed and how it works?
retry to scatter the region if the status is not `RUNNING` or `SUCCESS`.
### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by lance6716 — COMMENTED** [review]
it's better to let BR and lightning use same function

File: br/pkg/restore/split.go [hunk 1]
```diff
@@ -144,21 +144,14 @@ SplitRegions:
...
-	if scatterCount == len(scatterRegions) {
+
+	leftCnt := rs.waitForScatterRegions(ctx, scatterRegions, split.ScatterWaitUpperInterval)
+	if leftCnt == 0 {
 		log.Info("waiting for scattering regions done",
 			zap.Int("regions", len(scatterRegions)), zap.Duration("take", time.Since(startTime)))
 	} else {
 		log.Warn("waiting for scattering regions timeout",
-			zap.Int("scatterCount", scatterCount),
+			zap.Int("scatterCount", len(scatterRegions)-leftCnt),
```
Comment (3pointer) [inline br/pkg/restore/split.go:154]: maybe split the entries in log.

File: br/pkg/restore/split.go [hunk 2]
```diff
@@ -188,26 +181,44 @@ func (rs *RegionSplitter) hasHealthyRegion(ctx context.Context, regionID uint64)
...
+// if the latest operator is not `scatter-operator`, or its status is SUCCESS, it's likely that the
+// scatter region operator is finished.
+//
+// if the latest operator is `scatter-operator` and its status is TIMEOUT or CANCEL, the needRescatter
+// is true and the function caller needs to scatter this region again.
+func (rs *RegionSplitter) isScatterRegionFinished(ctx context.Context, regionID uint64) (bool, bool, error) {
 	resp, err := rs.client.GetOperator(ctx, regionID)
 	if err != nil {
-		return false, errors.Trace(err)
+		return false, false, errors.Trace(err)
```
Comment (3pointer) [inline br/pkg/restore/split.go:196]: why not retry here?

File: br/pkg/restore/split.go [hunk 3]
```diff
@@ -188,26 +181,44 @@ func (rs *RegionSplitter) hasHealthyRegion(ctx context.Context, regionID uint64)
 	return len(regionInfo.PendingPeers) == 0, nil
 }
 
-func (rs *RegionSplitter) isScatterRegionFinished(ctx context.Context, regionID uint64) (bool, error) {
+// isScatterRegionFinished check the latest successful operator and return the follow status:
+//
+//	return (finished, needRescatter, error)
+//
+// if the latest operator is not `scatter-operator`, or its status is SUCCESS, it's likely that the
```
Comment (3pointer) [inline br/pkg/restore/split.go:188]: if the latest operator is not `scatter-operator`
---

Will it become a mistake when the scatter-operator canceller and another operator succeed?

Comment (Leavrth) [reply br/pkg/restore/split.go:190]: yes. But it is ignored usually.

**Review by 3pointer — APPROVED** [review]

**Review by BornChanger — APPROVED** [review]

File: br/pkg/restore/split_test.go [hunk 4]
```diff
@@ -337,6 +341,114 @@ func TestSplitAndScatter(t *testing.T) {
...
+		client.InstallBatchScatterSupport()
+		runWaitScatter(t, client)
+	})
+}
+
+func TestXXX(t *testing.T) {
+	client := initTestClient(false)
+	client.InstallBatchScatterSupport()
+	runWaitScatter(t, client)
+}
```
[the remark this answers is not in the document]
Comment (Leavrth) [reply br/pkg/restore/split_test.go:355]: This is just a test to test the unit test, which I forget to remove...

**Pull request merged**